federated_aws_cli/login.py

```diff
@@ -3,111 +3,176 @@
-    code_challenge = generate_challenge(code_verifier)
-    state = base64_without_padding(os.urandom(32))
-
-    port = listener.get_available_port()
-    redirect_uri = "http://localhost:{}/redirect_uri".format(port)
-
-    url_parameters = {
-        "scope": scope,
-        "response_type": "code",
-        "redirect_uri": redirect_uri,
-        "client_id": client_id,
-        "code_challenge": code_challenge,
-        "code_challenge_method": "S256",
-        "state": state,
-    }
-    # We don't set audience here because Auth0 will set the audience on it's
-    # own
-    url = "{}?{}".format(authorization_endpoint, urlencode(url_parameters))
-
-    # Open the browser window to the login url
-    # Start the listener
-    logger.debug("About to spawn browser window to {}".format(url))
-    webbrowser.get("firefox").open(
-        url
-    )  # This specifies firefox to work around webbrowser.BackgroundBrowser sending stdout/stderr to the console :
-    # https://github.com/python/cpython/blob/783b794a5e6ea3bbbaba45a18b9e03ac322b3bd4/Lib/webbrowser.py#L177-L181
-    logger.debug("About to begin listener on port {}".format(port))
-    code, response_state, error_message = listener.get_code(port)
-
-    if code is None:
-        print("Something wrong happened, could not retrieve session data")
-        exit(1)
-
-    if state != response_state:
-        print("Error: State returned from IdP doesn't match state sent")
-        exit(1)
-
-    if error_message is not None:
-        print("An error occurred:")
-        print(error_message)
-        exit(1)
-
-    # Exchange the code for a token
-    headers = {"Content-Type": "application/json"}
-    body = {
-        "grant_type": "authorization_code",
-        "client_id": client_id,
-        "code_verifier": code_verifier,
-        "code": code,
-        "redirect_uri": redirect_uri,
-    }
-    r = requests.post(token_endpoint, headers=headers, data=json.dumps(body))
-    data = json.loads(r.text)
-
-    return data
-
-
-def main():
-    print(login())
-
-
-if __name__ == "__main__":
-    main()
+requests_cache.install_cache("{}/federated_aws_cli_cache".format(XDG_CACHE_HOME))
+
+
+class PkceLogin:
+    def __init__(self, well_known_url, client_id=None, scope="openid", port=None):
+        """
+        :param authorization_endpoint: URL of the OIDC authorization endpoint obtained from the discovery document
+        :param token_endpoint: URL of the OIDC token endpoint obtained from the discovery document
+        :param client_id: OIDC client_id of the native OIDC application
+        :param scope: OIDC scopes of claims to request
+        """
+        self.well_known_url = well_known_url
+        self.client_id = client_id
+        self.scope = scope
+        self.tokens = None
+        self.port = port if port is not None else listener.get_available_port()
+        self.redirect_uri = "http://localhost:{}/redirect_uri".format(self.port)
+
+    def __deferred_init__(self):
```

Comment:
I think `__deferred_init__` isn't a python reserved method. Were you trying to make a private method (e.g. `_deferred_init`) or enable [name mangling](https://docs.python.org/3/tutorial/classes.html#private-variables) (e.g. `__deferred_init`)?

Reply:
`_deferred_init` should do